### PR TITLE
(PUP-10713) Ensure environment cache is consistent when entries are cleared

### PR DESCRIPTION
**Pull request by joshcooper**
The environment cache has two methods to clear a cached environment (`clear_all` and `clear(name)`) and two methods to clear expired environment(s) (`clear_all_expired` and `clear_if_expired(name)`). An environment can become expired if the environment timeout is exceeded or if it is marked as expired through the puppetserver [environment-cache REST endpoint](https://puppet.com/docs/puppetserver/5.3/admin-api/v1/environment-cache.html#delete-puppet-admin-apiv1environment-cache).

When an environment is cleared, we need to delete the environment from memory (which contains the AST for parsed puppet code), notify the cache expiration service, delete the environment-specific text domain and environment-specific settings. Prior to this PR, we didn't clear environment consistently. To remedy that, this PR:

* Creates a `clear_entry` method to handle all cache removals. It mirrors the `add_entry` method.
* All `clear*` methods call `clear_entry`.

Each cache entry defined an `expires` and `expired?` method which basically did the same thing. The former returns the time when the entry expires and the latter returns true if the entry is currently expired. This PR removes the `expires` logic.

The `clear_all_expired` method tried to short-circuit if the `next expiration` time was in the future, see 299f5d597b0d1c32bf3c49adf125ba5d547351c4. Specifically:

> To avoid a worst case where there are many cached environments expiring
> a long time in the future, the algorithm keeps a sorted set of times
> where an expiration will take place. It uses this, and advice from the
> external cache service to determine if environments should be evicted or
> not. As environments are frequntly requested, this ensures that the list
> of cached environments is not traversed and checked when there is a
> guarnatee that there are no expirations.

However, the `next_expiration` time was not updated when `get_conf` or `clear` removed an expired environment. And even if that time was updated correctly, we can't get around the fact that an environment may be expired via the REST endpoint at any time. So we still need to ask the cache expiration service if an environment that hasn't reached its timeout was expired through the REST endpoint.

### CODE REVIEW

**Comment by joshcooper** [comment]
Supersedes https://github.com/puppetlabs/puppet/pull/8417

**Comment by joshcooper** [comment]
jenkins please test with servertests

**Comment by joshcooper** [comment]
jenkins please test this with servertests

**Comment by joshcooper** [comment]
One problem with the existing eviction logic (before and after this PR) is that Puppetserver calls `get!` for each `puppet/v3` REST request (eg node, catalog, file_metadata(s), report), and each call causes the cache to check the eviction status of N other environments. So if puppetserver is handling large numbers of environments, then we'll get quadratic behavior.

I think it'd be better if the environment cache kept a list of environments sorted on expiration time (in ascending order) and if puppetserver called directly into the cache to mark expired environments, rather than waiting for puppet to callback later. Whenever an environment is touched, we just need to remove it from the sorted list and add it back with the new expiration time.

Cache eviction then becomes "evict all marked environments" and "evict environments from the sorted list, and stop when we reach the first non-expired environment or end of list". We can also optimize for the case where the marked set is empty and the first entry in the expiration list is in the future.

I'll file that as a separate enhancement.

**Review by GabrielNagy — APPROVED** [review]

File: lib/puppet/environments.rb [hunk 1]
```diff
@@ -389,28 +382,38 @@ def get(name)
...
+
     # Clears the cache of the environment with the given name.
     # (The intention is that this could be used from a MANUAL cache eviction command (TBD)
     def clear(name)
-      @cache.delete(name)
-      Puppet::GettextConfig.delete_text_domain(name)
+      entry = @cache[name]
+      if entry
+        clear_entry(name, entry)
+      end
```
Comment (GabrielNagy) [inline lib/puppet/environments.rb:404]: can also write this on a single line

Comment (joshcooper) [reply lib/puppet/environments.rb]: Updated

**Pull request merged**